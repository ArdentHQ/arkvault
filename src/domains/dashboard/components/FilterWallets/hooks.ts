import { Networks } from "@ardenthq/sdk";
import { Contracts } from "@ardenthq/sdk-profiles";
import { useMemo } from "react";

import { FilterWalletsHookProperties } from "./FilterWallets.contracts";
import { useWalletConfig } from "@/domains/wallet/hooks";
import { useNetworks } from "@/app/hooks";

export const useWalletFilters = ({ profile }: { profile: Contracts.IProfile }) => {
	const { defaultConfiguration, setValue, walletsDisplayType, selectedNetworkIds, viewType } = useWalletConfig({
		profile,
	});

	const basicNetworks = useNetworks({ profile });
	const allWalletsLength = profile.wallets().values().length;
	const networks = useMemo<{ network: Networks.Network; isSelected: boolean }[]>(
		() =>
			basicNetworks.map((network) => ({
				isSelected: selectedNetworkIds.includes(network.id()),
				network,
			})),
		// eslint-disable-next-line react-hooks/exhaustive-deps
		[basicNetworks, selectedNetworkIds, allWalletsLength],
	);

	const isFilterChanged = useMemo(() => {
		if (walletsDisplayType !== defaultConfiguration.walletsDisplayType) {
			return true;
		}

		if (selectedNetworkIds.length < defaultConfiguration.selectedNetworkIds.length) {
			return true;
		}

		return false;
	}, [walletsDisplayType, selectedNetworkIds, defaultConfiguration]);

	return useMemo<FilterWalletsHookProperties>(
		() => ({
			defaultConfiguration,
			disabled: !profile.wallets().count(),
			isFilterChanged,
			networks,
			selectedNetworkIds,
			update: setValue,
			viewType,
			walletsDisplayType,
		}),
		[
			walletsDisplayType,
			selectedNetworkIds,
			viewType,
			isFilterChanged,
			networks,
			profile,
			setValue,
			defaultConfiguration,
		],
	);
};
