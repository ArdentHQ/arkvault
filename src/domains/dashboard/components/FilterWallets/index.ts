export * from "./FilterWallets";
export * from "./FilterWallets.contracts";
export * from "./hooks";
