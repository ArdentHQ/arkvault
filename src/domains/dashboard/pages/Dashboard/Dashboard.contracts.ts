export interface DashboardConfiguration {
	walletsDisplayType: "all" | "starred" | "ledger";
	selectedNetworkIds: string[];
}
