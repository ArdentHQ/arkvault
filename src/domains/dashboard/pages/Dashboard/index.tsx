export { Dashboard as default } from "./Dashboard";
export * from "./Dashboard.contracts";
