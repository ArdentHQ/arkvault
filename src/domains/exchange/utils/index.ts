export * from "./assertions";
export * from "./errors";
