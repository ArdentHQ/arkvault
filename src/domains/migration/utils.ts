import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
