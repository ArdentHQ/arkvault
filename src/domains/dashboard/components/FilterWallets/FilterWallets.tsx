import cn from "classnames";
import React from "react";
import { useTranslation } from "react-i18next";

import { FilterWalletsProperties } from "./FilterWallets.contracts";
import { Dropdown } from "@/app/components/Dropdown/Dropdown";
import { FilterNetworks } from "@/app/components/FilterNetwork";
import { Icon } from "@/app/components/Icon";

export const FilterWallets = ({ networks, walletsDisplayType, onChange }: FilterWalletsProperties) => {
	const { t } = useTranslation();

	const walletDisplayOptions = [
		{ label: t("COMMON.ALL"), value: "all" },
		{ label: t("COMMON.STARRED"), value: "starred" },
		{ label: t("COMMON.LEDGER"), value: "ledger" },
	];

	return (
		<div className="flex flex-col text-left" data-testid="FilterWallets">
			<div className="mb-6">
				<div className="font-semibold text-theme-secondary-text">
					{t("DASHBOARD.FILTER_WALLETS.CRYPTOASSET.TITLE")}
				</div>
				<div className="mt-1 text-sm text-theme-secondary-500 font-semibold">
					{t("DASHBOARD.FILTER_WALLETS.CRYPTOASSET.DESCRIPTION")}
				</div>
			</div>

			<FilterNetworks
				options={networks}
				onChange={(_: any, options: any[]) => {
					onChange?.(
						"selectedNetworkIds",
						options.filter((option) => option.isSelected).map((option) => option.network.id()),
					);
				}}
			/>

			<div className="mb-6 mt-2 border-t border-dotted border-theme-secondary-300 dark:border-theme-secondary-800" />

			<div className="flex flex-col">
				<div className="flex items-center justify-between">
					<div>
						<div className="font-semibold text-theme-secondary-text">
							{t("DASHBOARD.FILTER_WALLETS.WALLETS.TITLE")}
						</div>

						<div className="mt-2 pr-12 text-sm text-theme-secondary-500 font-semibold">
							{t("DASHBOARD.FILTER_WALLETS.WALLETS.DESCRIPTION")}
						</div>
					</div>
				
					<Dropdown
						dropdownClass="mx-4 sm:mx-0"
						toggleIcon="ChevronDownSmall"
						toggleSize="sm"
						options={walletDisplayOptions}
						onSelect={({ value }) => onChange?.("walletsDisplayType", value)}
						toggleContent={(isOpen: boolean) => (
							<div
								data-testid="filter-wallets__wallets"
								className="flex cursor-pointer items-center justify-end text-theme-secondary-text"
							>
								<span className="mr-2 inline-block font-semibold">
									{walletDisplayOptions.find((option) => option.value === walletsDisplayType)?.label}
								</span>
								<Icon
									name="ChevronDownSmall"
									className={cn("transition-transform", { "rotate-180": isOpen })}
									size="sm"
								/>
							</div>
						)}
					/>
				</div>

	
			</div>
		</div>
	);
};
