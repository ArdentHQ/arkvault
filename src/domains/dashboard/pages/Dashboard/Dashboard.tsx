import { uniq } from "@ardenthq/sdk-helpers";
import { Contracts, DTO } from "@ardenthq/sdk-profiles";
import React, { useMemo, useState } from "react";
import { useTranslation } from "react-i18next";
import { EmptyBlock } from "@/app/components/EmptyBlock";
import { useWalletFilters } from "@/domains/dashboard/components/FilterWallets";
import { PortfolioBreakdown } from "@/domains/dashboard/components/PortfolioBreakdown";
import { PortfolioHeader } from "@/domains/wallet/components/PortfolioHeader";
import { WalletsGroupsList } from "@/domains/wallet/components/WalletsGroup";
import { useLatestTransactions } from "@/domains/dashboard/hooks/use-latest-transactions";
import { TransactionTable } from "@/domains/transaction/components/TransactionTable";
import { TableWrapper } from "@/app/components/Table/TableWrapper";

import { Page, Section } from "@/app/components/Layout";
import { useConfiguration, useEnvironmentContext } from "@/app/contexts";
import { useActiveProfile } from "@/app/hooks";
import { WelcomeModal } from "@/domains/profile/components/WelcomeModal";
import { TransactionDetailModal } from "@/domains/transaction/components/TransactionDetailModal";

export const Dashboard: React.VFC = () => {
	const activeProfile = useActiveProfile();
	const wallet = activeProfile.wallets().first();
	const { t } = useTranslation();
	const { env } = useEnvironmentContext();
	const { profileIsSyncing, profileIsSyncingExchangeRates } = useConfiguration();
	const profileIsSyncedWithNetwork = !activeProfile.hasBeenPartiallyRestored();
	const walletsCount = activeProfile.wallets().count();
	const { selectedNetworkIds } = useWalletFilters({ profile: activeProfile });
	const [transactionModalItem, setTransactionModalItem] = useState<DTO.ExtendedConfirmedTransactionData | undefined>(
		undefined,
	);
	const showTransactions = activeProfile.appearance().get("dashboardTransactionHistory");
	const exchangeCurrency = activeProfile.settings().get<string>(Contracts.ProfileSetting.ExchangeCurrency);
	const { isLoadingTransactions, latestTransactions } = useLatestTransactions({
		profile: activeProfile,
		profileIsSyncing,
	});

	const liveNetworkIds = useMemo(
		() =>
			uniq(
				activeProfile
					.wallets()
					.values()
					.filter((wallet) => wallet.network().isLive())
					.map((wallet) => wallet.networkId()),
			),
		[activeProfile, walletsCount, profileIsSyncedWithNetwork], // eslint-disable-line react-hooks/exhaustive-deps
	);

	return (
		<>
			<Page pageTitle={t("DASHBOARD.DASHBOARD_PAGE.TITLE")} isBackDisabled={true}>
				<Section>
					<PortfolioHeader />
					<PortfolioBreakdown
						profile={activeProfile}
						profileIsSyncingExchangeRates={profileIsSyncingExchangeRates}
						liveNetworkIds={liveNetworkIds}
						selectedNetworkIds={selectedNetworkIds}
					/>
				</Section>

				<div className="lg:container md:px-10 lg:mx-auto">
					<WalletsGroupsList />
				</div>

				{showTransactions && (
					<Section className="mt-4 md:mt-0" data-testid="dashboard__transactions-view">
						<h2 className="mb-6 text-2xl font-bold">{t("DASHBOARD.LATEST_TRANSACTIONS.TITLE")}</h2>

						<TableWrapper className="overflow-hidden">
							<>

								<TransactionTable
									transactions={latestTransactions}
									exchangeCurrency={exchangeCurrency}
									hideHeader={!isLoadingTransactions && latestTransactions.length === 0}
									isLoading={isLoadingTransactions && latestTransactions.length === 0}
									skeletonRowsLimit={8}
									onRowClick={setTransactionModalItem}
									profile={activeProfile}
									coinName={wallet.currency()}
								/>

								{latestTransactions.length === 0 && !isLoadingTransactions && (
									<EmptyBlock>{t("DASHBOARD.LATEST_TRANSACTIONS.EMPTY_MESSAGE")}</EmptyBlock>
								)}
							</>

						</TableWrapper>
					</Section>
				)}
			</Page>

			<WelcomeModal profile={activeProfile} environment={env} />

			{transactionModalItem && (
				<TransactionDetailModal
					isOpen={!!transactionModalItem}
					transactionItem={transactionModalItem}
					profile={activeProfile}
					onClose={() => setTransactionModalItem(undefined)}
				/>
			)}
		</>
	);
};
