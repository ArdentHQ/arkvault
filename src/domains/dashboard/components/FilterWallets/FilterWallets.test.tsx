import { Contracts } from "@ardenthq/sdk-profiles";
import userEvent from "@testing-library/user-event";
import React from "react";

import { FilterOption } from "@/app/components/FilterNetwork";
import { DashboardConfiguration } from "@/domains/dashboard/pages/Dashboard";
import { env, getDefaultProfileId, render, screen } from "@/utils/testing-library";

import { FilterWallets } from "./FilterWallets";

let profile: Contracts.IProfile;
let networkOptions: FilterOption[];

const defaultConfiguration: DashboardConfiguration = {
	selectedNetworkIds: [],
	walletsDisplayType: "all",
};

describe("FilterWallets", () => {
	beforeAll(() => {
		profile = env.profiles().findById(getDefaultProfileId());

		const networks: Record<string, FilterOption> = {};

		for (const wallet of profile.wallets().values()) {
			const networkId = wallet.networkId();

			if (!networks[networkId]) {
				networks[networkId] = {
					isSelected: false,
					network: wallet.network(),
				};
			}
		}

		networkOptions = Object.values(networks);
	});

	it("should render", () => {
		const { container } = render(
			<FilterWallets selectedNetworkIds={[]} defaultConfiguration={defaultConfiguration} />,
		);

		expect(container).toMatchSnapshot();
	});

	it("should render with networks selection", () => {
		const { container } = render(
			<FilterWallets
				selectedNetworkIds={[]}
				networks={networkOptions}
				defaultConfiguration={defaultConfiguration}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it("should emit onChange for network selection", async () => {
		const onChange = vi.fn();

		render(
			<FilterWallets
				selectedNetworkIds={[]}
				networks={networkOptions}
				onChange={onChange}
				defaultConfiguration={defaultConfiguration}
			/>,
		);

		await userEvent.click(screen.getByTestId(`NetworkOption__${networkOptions[0].network.id()}`));

		expect(onChange).toHaveBeenCalledWith("selectedNetworkIds", [networkOptions[0].network.id()]);
	});

	it("should emit onChange for wallets display type change", async () => {
		const onChange = vi.fn();

		render(
			<FilterWallets
				selectedNetworkIds={[]}
				networks={networkOptions}
				onChange={onChange}
				defaultConfiguration={defaultConfiguration}
			/>,
		);

		await userEvent.click(screen.getByTestId("filter-wallets__wallets"));

		await userEvent.click(screen.getByTestId("dropdown__option--0"));

		expect(onChange).toHaveBeenCalledWith("walletsDisplayType", "all");
	});

	it("should not emit onChange for wallet display type change", async () => {
		const onChange = vi.fn();

		render(
			<FilterWallets
				selectedNetworkIds={[]}
				networks={networkOptions}
				defaultConfiguration={defaultConfiguration}
			/>,
		);

		await userEvent.click(screen.getByTestId("filter-wallets__wallets"));

		await userEvent.click(screen.getByTestId("dropdown__option--0"));

		expect(onChange).not.toHaveBeenCalled();
	});
});
