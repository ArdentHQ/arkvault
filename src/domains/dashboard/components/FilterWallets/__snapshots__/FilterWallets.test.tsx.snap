// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterWallets should render 1`] = `
<div>
  <div
    class="flex flex-col text-left"
    data-testid="FilterWallets"
  >
    <div
      class="mb-5"
    >
      <div
        class="font-semibold text-theme-secondary-text"
      >
        Filter Networks
      </div>
      <div
        class="mt-1 text-sm text-theme-secondary-500"
      >
        Select the networks to show.
      </div>
    </div>
    <div
      class="space-y-4"
    />
    <div
      class="mt-6 mb-8 border-t border-dotted border-theme-secondary-300 dark:border-theme-secondary-800"
    />
    <div
      class="flex flex-col"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="font-semibold text-theme-secondary-text"
        >
          Your Wallets
        </div>
        <div
          class="static sm:relative"
        >
          <span
            data-testid="dropdown__toggle"
          >
            <div
              class="flex cursor-pointer items-center justify-end text-theme-secondary-text"
              data-testid="filter-wallets__wallets"
            >
              <span
                class="mr-2 inline-block font-semibold"
              />
              <div
                class="transition-transform css-15txs7d"
                height="10"
                width="10"
              >
                <svg>
                  chevron-down-small.svg
                </svg>
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        class="mt-1 pr-12 text-sm text-theme-secondary-500"
      >
        Select which wallets to view.
      </div>
    </div>
  </div>
</div>
`;

exports[`FilterWallets should render with networks selection 1`] = `
<div>
  <div
    class="flex flex-col text-left"
    data-testid="FilterWallets"
  >
    <div
      class="mb-5"
    >
      <div
        class="font-semibold text-theme-secondary-text"
      >
        Filter Networks
      </div>
      <div
        class="mt-1 text-sm text-theme-secondary-500"
      >
        Select the networks to show.
      </div>
    </div>
    <div
      class="space-y-4"
    >
      <div
        data-testid="FilterNetwork"
      >
        <div
          class="mb-3 text-sm font-bold text-theme-secondary-400"
        >
          Development Networks
        </div>
        <ul
          class="flex flex-wrap"
          data-testid="NetworkOptions"
        >
          <li
            class="inline-block cursor-pointer pr-5 pb-5"
            data-testid="NetworkOption__ark.devnet"
          >
            <div
              class="relative border-theme-secondary-300 dark:border-theme-secondary-800 ring-theme-background duration-100 css-14jhlj1"
            >
              <div
                aria-label="ARK Devnet"
                class="inline-flex h-5 w-5 items-center justify-center text-theme-secondary-700 border-theme-secondary-300 dark:border-theme-secondary-700"
                data-testid="NetworkIcon-ARK-ark.devnet"
              >
                <div
                  class="text-theme-secondary-300 css-1i4b0eq"
                  data-testid="NetworkIcon__icon"
                  height="20"
                  width="20"
                >
                  <svg>
                    ark.svg
                  </svg>
                </div>
              </div>
              <span
                class="flex border-2 rounded-full justify-center items-center align-middle bg-theme-background border-theme-secondary-300 dark:border-theme-secondary-800 css-19jfgr"
              >
                <span />
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="mt-6 mb-8 border-t border-dotted border-theme-secondary-300 dark:border-theme-secondary-800"
    />
    <div
      class="flex flex-col"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="font-semibold text-theme-secondary-text"
        >
          Your Wallets
        </div>
        <div
          class="static sm:relative"
        >
          <span
            data-testid="dropdown__toggle"
          >
            <div
              class="flex cursor-pointer items-center justify-end text-theme-secondary-text"
              data-testid="filter-wallets__wallets"
            >
              <span
                class="mr-2 inline-block font-semibold"
              />
              <div
                class="transition-transform css-15txs7d"
                height="10"
                width="10"
              >
                <svg>
                  chevron-down-small.svg
                </svg>
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        class="mt-1 pr-12 text-sm text-theme-secondary-500"
      >
        Select which wallets to view.
      </div>
    </div>
  </div>
</div>
`;
